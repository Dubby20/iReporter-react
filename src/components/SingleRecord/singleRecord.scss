@import '../../../public/styles/color.scss';

.error {
  margin: 8px;
}

.edit-btn {
  color: $main-color;
  float: right;
  font-weight: 600;
  cursor: pointer;
  border: none;
  background: none;
  outline: none;
}

.hidden {
  display: none;
}

.locate {
  margin: 8px 0px;
}

#input-location {
  width: 25%;
  display: block;
  margin-right: auto;
  margin-left: auto;
}

.c-btn {
  font-size: 13px;
  height: 28px;
  min-width: 56px;
  padding: 0 12px 1px;
  text-decoration: none;
  outline: none;
  cursor: pointer;
  border: none;
  border-radius: 4px;
  align-items: center;
  position: relative;
  display: inline-flex;
  justify-content: center;
  text-align: center;
  white-space: nowrap;
  margin: 8px 0px;
}

.outline {
  transition: all 80ms linear;
  background: $text-color;
  box-shadow: inset 0 0 0 1px rgba(29, 28, 29, 0.3);
  color: #1d1c1d;
  font-weight: 700;
}

.primary {
  margin-left: 80px;
  transition: all 80ms linear;
  background: #007a5a;
  color: $text-color;
  font-weight: 900;
  box-shadow: none;
}

.edit-btn {
  color: $main-color;
  float: right;
  font-weight: 600;
  cursor: pointer;
  border: none;
  background: none;
  outline: none;
}

.edit-btn:hover {
  color: #007a5a;
}



.image-layout,
.video-layout {
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  width: 100%;
  /* border: 1px solid black; */
}

.image-layout li,
.video-layout li {
  margin: 12px;
}

img {
  vertical-align: middle;
}

.delete-record {
  margin-top: 60px;
  text-align: center;
}

.fas {
  display: inline-block;
  
}

.fa-trash {
  color: #5b616a;
  outline: none;
  
}

.fa-color {
  color: $primary-color;
}

.trash {
  border: none;
  cursor: pointer;
  

  & > * {
    pointer-events: none;
  }
}

.trash:hover {
  color: $primary-color;
}

.del-content {
  border-radius: 8px;
  background: $text-color;
  width: 500px;
  margin-right: auto;
  margin-left: auto;
}

.animate {
  -webkit-animation: animatezoom 0.6s;
  animation: animatezoom 0.6s;
}

@keyframes modalopen {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

.close-btn,
.close-page {
  top: 0;
  color: #000;
  float: right;
  font-size: 30px;
  font-weight: bold;
  padding-right: 8px;
}

.close-page:hover,
.close-page:focus {
  color: $primary-color;
  text-decoration: none;
  cursor: pointer;
}

.title-background {
  background: #e4ebf1;
  padding: 20px;
  line-height: 1.25;
  border-radius: 8px 8px 0px 0px;
}

.del-question {
  line-height: 18px;
  padding: 20px;
}

.display-btn {
  padding: 1rem 20px;
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  flex-wrap: nowrap;
}

.bg-red {
  background: $primary-color;
  color: text-color;
  border-radius: 4px;
  cursor: pointer;
  border: none;
  font-size: 14px;
  line-height: 18px;
  padding: 10px 16px;
  outline: none;
}

.bg-cancel {
  background: #e4ebf1;
  color: #5a616a;
  border-radius: 4px;
  cursor: pointer;
  border: none;
  font-size: 14px;
  line-height: 18px;
  padding: 10px 16px;
  outline: none;
}

.main-content {
  display: flex;
  flex-direction: column;
  align-items: center;

  & > * {
    margin: 0.5rem 0;
    font-size: 1.2rem;
  }
}

.actions {
  display: inline-flex;
  flex-direction: column;
  align-items: flex-end;
  position: absolute;
  right: 30px;
  top: 35%;

  & > * {
    margin: 0.5rem 0;
    font-size: 1.2rem;
  }
}

.delete-record {
  margin-top: 60px;
}