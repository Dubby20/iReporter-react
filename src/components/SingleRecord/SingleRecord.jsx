/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable react/no-array-index-key */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Loader from '../Loader/Loader';
import { singleRecordRequest } from '../../actions/singleRecordActions';
import editCommentRequest from '../../actions/editcommentActions';
import deleteRecordRequest from '../../actions/deleteRecordActions';
import Comment from '../Comment/Comment';
import './singleRecord.scss';



export class SingleRecord extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isVisible: false,
      commentInput: '',
      showModal: false
    };
  }

  async componentDidMount() {

    const {
      match: {
        params: { id, type }
      },
      singleRecordRequest
    } = this.props;

    await singleRecordRequest({ id, type });
  }

  renderImage = (image = []) => {
    if (image.length === 0) {
      return 'No Image Uploaded';
    }
    const displayImage = image.map(
      (img, i) =>
        <img src={img} key={i} alt="" className="item" height="200" width="240" />

    );
    return displayImage;
  };

  renderVideo = (video = []) => {
    if (video.length === 0) {
      return 'No Video Uploaded';
    }
    const displayVideo = video.map(
      (vid, i) =>
        <video key={i} width="240" height="180" controls><source src={vid} /></video>

    );
    return displayVideo;
  };

  toggleCommentInput = () => this.setState((prevState) => ({
    isVisible: !prevState.isVisible,
  }))

  toggleModal = () => this.setState((prevState) => ({
    showModal: !prevState.showModal,
  }))


  handleClick = async (event) => {
    const { target } = event;
    const { match: {
      params: { id, type }
    }, editComment, deleteRecord, history } = this.props;
    const { commentInput } = this.state;

    switch (target.name) {
      case 'cancel':
        this.toggleCommentInput();
        break;
      case 'save':
        await editComment({ id, type, newComment: commentInput });
        this.toggleCommentInput();
        break;
      case 'delete':
        await deleteRecord(id);
        if (type === 'red-flag') {
          history.push('/redFlag');
        } else {
          history.push('/intervention');
        }
        break;
      default:
        this.toggleCommentInput();
    }
  }



  handleChange = (event) => {
    this.setState({
      commentInput: event.target.value,
    });
  }

  render() {
    const { isLoading, record, user } = this.props;
    const { isVisible, showModal } = this.state;

    if (isLoading) {
      return <div className="center"><Loader /></div>;
    }

    return (
      <div className="record-container">
        <div className="main-content">
          <p className="type">
            {`Type: ${record.type}`}
          </p>
          <p className="status-p">
            {`Status: ${record.status}`}
          </p>
          <p className="location">
            {`Location: ${record.location}`}
          </p>
          {!isVisible && (
            <p id="comment">
              {record.comment}
            </p>
          )}
        </div>
        {
          user.id === record.user_id && (
            <div className="actions">
              <button className="edit-btn change-location">
                Change location
              </button>
              <button
                name="edit"
                type="submit"
                className="edit-btn action-btn comment-btn"
                onClick={this.handleClick}
              >
                Edit Comment
              </button>
            </div>
          )
        }
        <div className="comment-div">
          {isVisible && (
            <Comment
              initialValue={record.comment}
              handleClick={this.handleClick}
              handleChange={this.handleChange}
            />
          )}
        </div>
        <div className="error" />
        <div className="hidden">
          <div className="form-group locate">
            <input type="text" id="input-location" className="form-control" />
          </div>
          <button type="submit" className="c-btn outline">Cancel</button>
          <button type="submit" className="c-btn primary">Update Location</button>
        </div>

        <div id="image-frame">
          <ul className="image-layout">
            <li className="image-list">
              {this.renderImage(record.images)}
            </li>
          </ul>
        </div>
        <div className="video-frame">
          <ul className="video-layout">
            <li className="video-list">
              {this.renderVideo(record.videos)}
            </li>
          </ul>
        </div>
        {user.id === record.user_id && (
          <div className="delete-record action-btn">
            <button type="submit" className="trash" onClick={this.toggleModal}><i className="fas fa-trash fa-color fa-2x" /></button>
          </div>
        )
        }
        {showModal && (
          <div id="delete-modal" className="modal">
            <div className="del-content animate">
              <span onClick={this.toggleModal} className="close-page">&times;</span>
              <div className="del-body">
                <div className="title-background"><h4>Delete Record</h4></div>
                <p className="del-question">Are you sure you want to delete the selected record?</p>
                <hr />
                <div className="display-btn">
                  <button onClick={this.toggleModal} type="submit" className="bg-cancel">Cancel</button>
                  <button type="submit" name="delete" className="bg-red" onClick={this.handleClick}>Delete</button>
                </div>
              </div>
            </div>
          </div>
        )}
      </div>
    );
  }
}

SingleRecord.propTypes = {
  singleRecordRequest: PropTypes.func.isRequired,
};


const mapStateToProps = state => ({
  isLoading: state.recordReducer.isLoading,
  record: state.recordReducer.record,
  user: state.authReducer.user.user,
});

const mapDispatchToProps = dispatch => ({
  singleRecordRequest: payload => dispatch(singleRecordRequest(payload)),
  editComment: payload => dispatch(editCommentRequest(payload)),
  deleteRecord: id => dispatch(deleteRecordRequest(id))
});


const connectedSingleRecord = connect(mapStateToProps, mapDispatchToProps)(SingleRecord);

export default connectedSingleRecord;