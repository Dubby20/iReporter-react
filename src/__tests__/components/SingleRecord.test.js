import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { SingleRecord } from '../../components/SingleRecord/SingleRecord';

const mockStore = configureStore([thunk]);
let wrapper;

const props = {
  singleRecordRequest: jest.fn().mockReturnValue(() => { }),
  history: {
    push: jest.fn(),
    location: {
      pathname: '/'
    }
  },
  match: {
    params: {
      id: 1,
      type: "intervevention"
    }
  },
  editComment: jest.fn(),
  deleteRecord: jest.fn(),

  record: {
    id: 1,
    type: 'intervention',
    comment: "This is a comment",
    images: ["https://imgur.com"],
    videos: ["res.cloudinary"],
    location: "6.7868, 7.8993",
    status: 'draft',
    user_id: 87
  },

  user: {
    email: "jac@yahoo.com",
    password: '567dfghj',
    id: 87
  }
};




const store = mockStore({

  notifyReducer: {
    type: '',
    show: false,
    message: ''
  }
});

describe("SingleRecord component", () => {

  it('should render singleRecord page', () => {
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <SingleRecord {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find('.record-container').exists()).toBe(true);

  });

  it('should render loader', () => {
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <SingleRecord {...props} isLoading />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find('Loader').exists()).toBe(true);

  });

  it('should handleClick', (done) => {

    const event = {
      target: {
        name: 'comment',
        value: 'Hello world'
      }
    };


    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <SingleRecord
            {...props}
          />
        </MemoryRouter>
      </Provider>
    );
    const btn = wrapper.find('.comment-btn');
    btn.simulate('click');
    const comment = wrapper.find('textarea[name="comment"]');
    comment.simulate('change', event);

    expect(wrapper.find('SingleRecord').instance().state.commentInput).toEqual('Hello world');

    setTimeout(() => {
      done();
    }, 2000);
  });

  it('should handle cancel button', (done) => {

    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <SingleRecord
            {...props}
          />
        </MemoryRouter>
      </Provider>
    );
    const spy = jest.spyOn(wrapper.find('SingleRecord').instance(), 'toggleCommentInput');
    const btn = wrapper.find('.comment-btn');
    btn.simulate('click');

    const cancelButton = wrapper.find('.outline').at(0);
    cancelButton.simulate('click');
    expect(spy).toHaveBeenCalled();

    setTimeout(() => {
      done();
    }, 2000);
  });

  it('should handle save button', (done) => {

    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <SingleRecord
            {...props}
          />
        </MemoryRouter>
      </Provider>
    );
    const spy = jest.spyOn(wrapper.find('SingleRecord').instance(), 'toggleCommentInput');
    const btn = wrapper.find('.comment-btn');
    btn.simulate('click');

    const cancelButton = wrapper.find('.primary').at(0);
    cancelButton.simulate('click');
    expect(spy).toHaveBeenCalled();

    setTimeout(() => {
      done();
    }, 2000);
  });

  it('should handle delete button', (done) => {

    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <SingleRecord
            {...props}
          />
        </MemoryRouter>
      </Provider>
    );

    const btn = wrapper.find('.fa-trash');
    btn.simulate('click');

    const deleteButton = wrapper.find('.bg-red').at(0);
    deleteButton.simulate('click');


    setTimeout(() => {
      expect(props.history.push).toBeCalledWith('/intervention');
      done();
    }, 2000);
  });


  it('should handle render image when null', (done) => {
    const props = {
      singleRecordRequest: jest.fn().mockReturnValue(() => { }),
      history: {
        push: jest.fn(),
        location: {
          pathname: '/'
        }
      },
      match: {
        params: {
          id: 1,
          type: "intervevention"
        }
      },

      record: {
        id: 1,
        type: 'intervention',
        comment: "This is a comment",
        images: [],
        videos: ["res.cloudinary"],
        location: "6.7868, 7.8993",
        status: 'draft',
        user_id: 87
      },

      user: {
        email: "jac@yahoo.com",
        password: '567dfghj',
        id: 87
      }
    };

    const store = mockStore({
      notifyReducer: {
        type: '',
        message: '',
        show: true
      }
    });
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <SingleRecord
            {...props}
          />
        </MemoryRouter>
      </Provider>
    );


    expect(wrapper.find('.image-list').text()).toEqual('No Image Uploaded');

    setTimeout(() => {
      done();
    }, 2000);
  });


  it('should handle render video when null', (done) => {
    const props = {
      singleRecordRequest: jest.fn().mockReturnValue(() => { }),
      history: {
        push: jest.fn(),
        location: {
          pathname: '/'
        }
      },
      match: {
        params: {
          id: 1,
          type: "intervevention"
        }
      },

      record: {
        id: 1,
        type: 'intervention',
        comment: "This is a comment",
        images: [],
        videos: [],
        location: "6.7868, 7.8993",
        status: 'draft',
        user_id: 87
      },

      user: {
        email: "jac@yahoo.com",
        password: '567dfghj',
        id: 87
      }
    };

    const store = mockStore({
      notifyReducer: {
        type: '',
        message: '',
        show: true
      }
    });
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <SingleRecord
            {...props}
          />
        </MemoryRouter>
      </Provider>
    );


    expect(wrapper.find('.video-list').text()).toEqual('No Video Uploaded');

    setTimeout(() => {
      done();
    }, 2000);
  });

});