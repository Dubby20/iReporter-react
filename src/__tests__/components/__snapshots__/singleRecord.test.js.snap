// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleRecord component should render singleRecord page 1`] = `ReactWrapper {}`;
